File: services/notification-alerting-service/src/models/notification.rs
use chrono::{DateTime, Utc};
use event_bus_control::contracts::NotificationEvent as SharedNotificationEvent;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sqlx::FromRow;
use uuid::Uuid;

#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
pub struct NotificationRecord {
    pub id: Uuid,
    pub user_id: Option<Uuid>,
    pub title: String,
    pub body: String,
    pub category: String,
    pub severity: String,
    pub status: String,
    pub channels: Value,
    pub metadata: Value,
    pub created_at: DateTime<Utc>,
    pub read_at: Option<DateTime<Utc>>,
}

impl Default for NotificationRecord {
    fn default() -> Self {
        Self {
            id: Uuid::nil(),
            user_id: None,
            title: String::new(),
            body: String::new(),
            category: String::new(),
            severity: String::new(),
            status: String::new(),
            channels: Value::Null,
            metadata: Value::Null,
            created_at: DateTime::<Utc>::from_timestamp(0, 0).unwrap_or_else(Utc::now),
            read_at: None,
        }
    }
}

#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
pub struct NotificationDelivery {
    pub id: Uuid,
    pub notification_id: Uuid,
    pub channel: String,
    pub status: String,
    pub response: Option<String>,
    pub created_at: DateTime<Utc>,
}

pub type NotificationEvent = SharedNotificationEvent<NotificationRecord>;

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct SendNotificationRequest {
    pub user_id: Option<Uuid>,
    pub title: String,
    pub body: String,
    pub severity: Option<String>,
    pub category: Option<String>,
    pub channels: Option<Vec<String>>,
    pub metadata: Option<Value>,
}

#[derive(Debug, Deserialize)]
pub struct ListNotificationsQuery {
    pub status: Option<String>,
    pub limit: Option<i64>,
}

#[derive(Debug, Deserialize)]
pub struct WebSocketQuery {
    pub ticket: String,
}

#[derive(Debug, Serialize)]
pub struct WebSocketTicketResponse {
    pub ticket: String,
    pub expires_in: i64,
}

File: services/notification-alerting-service/internal/models/notification.go
// Package models holds the wire types for notification-alerting-service.
//
// JSON tag set + field types match the Rust crate verbatim so the
// frontend, edge-gateway and event-bus consumers do not see a wire
// difference between the two implementations.
package models

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// NotificationRecord is the row in `notifications` and the wire payload
// returned by the list / mark_read endpoints.
type NotificationRecord struct {
	ID        uuid.UUID       `json:"id"`
	UserID    *uuid.UUID      `json:"user_id"`
	Title     string          `json:"title"`
	Body      string          `json:"body"`
	Category  string          `json:"category"`
	Severity  string          `json:"severity"`
	Status    string          `json:"status"`
	Channels  json.RawMessage `json:"channels"`
	Metadata  json.RawMessage `json:"metadata"`
	CreatedAt time.Time       `json:"created_at"`
	ReadAt    *time.Time      `json:"read_at"`
}

// NotificationDelivery is the row in `notification_deliveries`.
type NotificationDelivery struct {
	ID             uuid.UUID `json:"id"`
	NotificationID uuid.UUID `json:"notification_id"`
	Channel        string    `json:"channel"`
	Status         string    `json:"status"`
	Response       *string   `json:"response,omitempty"`
	CreatedAt      time.Time `json:"created_at"`
}

// SendNotificationRequest is the body of POST /notifications/send and POST /internal/notifications.
type SendNotificationRequest struct {
	UserID   *uuid.UUID      `json:"user_id"`
	Title    string          `json:"title"`
	Body     string          `json:"body"`
	Severity *string         `json:"severity,omitempty"`
	Category *string         `json:"category,omitempty"`
	Channels []string        `json:"channels,omitempty"`
	Metadata json.RawMessage `json:"metadata,omitempty"`
}

// ListNotificationsQuery models the GET /notifications query string.
type ListNotificationsQuery struct {
	Status *string
	Limit  int64
}

// WebSocketTicketResponse is the response of POST /notifications/ws-ticket.
type WebSocketTicketResponse struct {
	Ticket    string `json:"ticket"`
	ExpiresIn int64  `json:"expires_in"`
}
